cluster/src/shard_lock_manager.rs

```diff
@@ -0,0 +1,505 @@
+                        return
+                    }
+                }
+            }
+        });
+
+        self.lease_check_handle = Some(handle);
+        self.lease_keepalive_stopper = Some(keepalive_stop_sender);
+        self.lease_id = Some(lease_id);
+        Ok(())
+    }
+}
+
+impl ShardLockManager {
+    pub fn new(
+        key_prefix: String,
+        node_name: String,
+        etcd_client: Client,
+        runtime: RuntimeRef,
+    ) -> ShardLockManager {
+        let value = Bytes::from(ShardLockValue { node_name }.encode_to_vec());
+
+        ShardLockManager {
+            key_prefix,
+            value,
+            lease_ttl_sec: DEFAULT_LEASE_SECS,
+            etcd_client,
+            runtime,
+            shard_locks: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    /// Grant lock to the shard.
+    ///
+    /// If the lock is already granted, return false. The `on_lock_expired` will
+    /// be called when the lock lease is expired, but it won't be triggered if
+    /// the lock is revoked.
+    pub async fn grant_lock<OnExpired, Fut>(
+        &self,
+        shard_id: u32,
+        on_lock_expired: OnExpired,
+    ) -> Result<bool>
+    where
+        OnExpired: FnOnce(ShardId) -> Fut + Send + 'static,
+        Fut: Future<Output = ()> + Send + 'static,
+    {
+        info!("Try to grant lock for shard:{shard_id}");
+        {
+            let shard_locks = self.shard_locks.read().await;
+            if shard_locks.contains_key(&shard_id) {
+                warn!("The lock has been granted, shard_id:{shard_id}");
+                return Ok(false);
+            }
+        }
+
+        // Double check whether the locks has been granted.
+        let mut shard_locks = self.shard_locks.write().await;
+        if shard_locks.contains_key(&shard_id) {
+            warn!("The lock has been granted, shard_id:{shard_id}");
+            return Ok(false);
+        }
+
+        let mut shard_lock = ShardLock::new(
+            shard_id,
+            &self.key_prefix,
+            self.value.clone(),
+            self.lease_ttl_sec,
+        );
+        let mut etcd_client = self.etcd_client.clone();
+        shard_lock
+            .grant(on_lock_expired, &mut etcd_client, &self.runtime)
+            .await?;
+
+        shard_locks.insert(shard_id, shard_lock);
+
+        info!("Finish granting lock for shard:{shard_id}");
+        Ok(true)
+    }
+
+    /// Revoke the shard lock.
+    ///
+    /// If the lock is not exist, return false. And the `on_lock_expired` won't
+    /// be triggered.
+    pub async fn revoke_lock(&self, shard_id: u32) -> Result<bool> {
+        info!("Try to revoke lock for shard:{shard_id}");
+
+        let mut shard_locks = self.shard_locks.write().await;
+        let shard_lock = shard_locks.remove(&shard_id);
+        match shard_lock {
+            Some(mut v) => {
+                let mut etcd_client = self.etcd_client.clone();
+                v.revoke(&mut etcd_client).await?;
```

Comment:
What if this revoke failed?

Reply:
The error will be thrown up to the CeresMeta finally. And in the ceresmeta, the close shard call will fail and the following operations will be cancelled.